Add v5 upgrade guide (#757)
bugsnag/bugsnag-android: UPGRADING.md

Upgrade from 4.X to 5.X
-----------------------

__This version contains many breaking changes__. It is part of an effort to unify our notifier libraries across platforms, making the user interface more consistent, and implementations better on multi-layered environments where multiple Bugsnag libraries need to work together (such as React Native).

# Interfaces

> ## Bugsnag

Author: This change feels inconsistent and potentially confusing as all the other titles are named after their class name. Perhaps there's some middle ground?

`Bugsnag.init()` has been renamed to `Bugsnag.start()`.

Many of the previous methods on `Bugsnag` should now be called on `Configuration` instead.

> For example, `maxBreadcrumbs` can now only be set on `Configuration` before `Bugsnag` is initialized:

Reviewer: ```suggestion
For example, `maxBreadcrumbs` can now only be set on `Configuration` when Bugsnag is started:
```

```kotlin
val config = Configuration("my-api-key")
config.maxBreadcrumbs = 35
Bugsnag.init(this, config)
```

The full list of altered methods and their intended replacements can be found below:

| v4.x API  | v5.x API |
| ------------- | ------------- |
| `Bugsnag#addToTab`  | `Bugsnag#addMetadata(String, String, Object)` |
| `Bugsnag#beforeNotify` | `Bugsnag#addOnError` |
| `Bugsnag#beforeRecordBreadcrumb` | `Bugsnag#addOnBreadcrumb` |
| `Bugsnag#clearBreadcrumbs` | `Bugsnag#addOnBreadcrumb/Configuration#setMaxBreadcrumbs` |
| `Bugsnag#clearTab` | `Bugsnag#clearMetadata(String)` |
| `Bugsnag#clearUser` | `Bugsnag#setUser` |
| `Bugsnag#disableExceptionHandler` | `Configuration#setEnabledErrorTypes` |

> | `Bugsnag#enableExceptionHandler` | `Configuration#setEnabledErrorTypes` |

Author: I believe `Configuration#setEnabledErrorTypes` is more correct here as this API allows for the exception handler to be switched off in isolation. I'll update the cell to mention `autoDetectErrors` as another alternative API.

| `Bugsnag#setAutoCaptureSessions` | `Configuration#setAutoTrackSessions` |
| `Bugsnag#setBuildUUID` | `Configuration#setBuildUuid` |
| `Bugsnag#setEndpoint` | `Configuration#setEndpoints` |
| `Bugsnag#setErrorReportApiClient` | `Configuration#setDelivery` |
| `Bugsnag#setFilters` | `Configuration#setRedactedKeys` |
| `Bugsnag#setIgnoreClasses` | `Configuration#setIgnoreClasses` |
| `Bugsnag#setLoggingEnabled` | `Configuration#setLogger` |
| `Bugsnag#setMaxBreadcrumbs` | `Configuration#setMaxBreadcrumbs` |
| `Bugsnag#setMetaData` | `Configuration#addMetadata` |
| `Bugsnag#setNotifyReleaseStages` | `Configuration#setEnabledReleaseStages` |
| `Bugsnag#setProjectPackages` | `Configuration#setProjectPackages` |
| `Bugsnag#setReleaseStage` | `Configuration#setReleaseStage` |
| `Bugsnag#setSendThreads` | `Configuration#setSendThreads` |
| `Bugsnag#setSessionTrackingApiClient` | `Configuration#setDelivery` |
| `Bugsnag#setUserEmail` | `Bugsnag#setUser` |
| `Bugsnag#setUserId` | `Bugsnag#setUser` |
| `Bugsnag#setUserName` | `Bugsnag#setUser` |

See the [full documentation](https://docs.bugsnag.com/platforms/android) for more information.

> ## Client

Author: I think we should keep this section because users of v4 will have been using `Client`.

I can update the suggested v5 API to point at the `Bugsnag` facade methods instead.

    event.context = "My Unique context"
    return true
}
```

If you use the NDK, you will need to implement your own native `on_error` callback _in addition_ to any JVM callbacks. This native `on_error` callback will run for fatal C/C++ errors only.

See the [full documentation](https://docs.bugsnag.com/platforms/android) for more information.

## Breadcrumb

| v4.x API  | v5.x API |
| ------------- | ------------- |
| `Breadcrumb#getName`  | `Breadcrumb#getMessage` |

## SessionTrackingPayload
`SessionTrackingPayload` is now called `SessionPayload`.

It is now possible to redact the `app` and `device` information captured on sessions via an `OnSessionCallback`.
See the [full documentation](https://docs.bugsnag.com/platforms/android) for more information.

## BeforeRecordBreadcrumb
`BeforeRecordBreadcrumb` is now called `OnBreadcrumbCallback`.

| v4.x API  | v5.x API |
| ------------- | ------------- |
| `BeforeRecordBreadcrumb#shouldRecord` | `OnBreadcrumbCallback#onBreadcrumb` |

## Delivery

| v4.x API  | v5.x API |
| ------------- | ------------- |
| `Delivery#deliver(Report, Configuration)` | `Delivery#deliver(SessionPayload, DeliveryParams)` |

> | `Delivery#deliver(SessionTrackingPayload, Configuration)` | `Delivery#deliver(Report, DeliveryParams)` |

Author: We need to pass the entire `Report` here rather than the `Event` as it contains all the information required for an Error Reporting API payload. I believe this will eventually be replaced by individual configuration options e.g. for setting a proxy
